File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int n = heights.length;
        int[] right = new int[n];
        int[] left = new int [n];
        Stack<Integer> st = new Stack<>();
        for(int i=0;i<n;i++){
            while(!st.isEmpty() && heights[st.peek()]>=heights[i])
                st.pop();
            if(st.isEmpty()) right[i] = -1;
            else right[i] = st.peek();
            st.push(i);
        }
        while(!st.isEmpty()) st.pop();
        for(int i=n-1;i>=0;i--){
            while(!st.isEmpty() && heights[st.peek()]>=heights[i])
                st.pop();
            if(st.isEmpty()) left[i] =n;
            else left[i] = st.peek();
            st.push(i); 
        }
        int maxArea = Integer.MIN_VALUE;
        for(int i=0;i<n;i++){
            int breadth = left[i]-right[i]-1;
            int area = breadth* heights[i];
            maxArea = Math.max(maxArea , area);
        }
        return maxArea;
    }
}

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n = heights.size();
        vector<int>right(n),left(n);
        stack<int>st;
        //left->right who is smaller than me one to my left 
        for(int i=0;i<n;i++){
            while(!st.empty() && heights[st.top()]>=heights[i])
                st.pop();
            if(st.empty()) right[i] = -1;
            else right[i] = st.top();
            st.push(i);
        }
        //clearing the stack 
        while(!st.empty()) st.pop();
        //right->left who is smaller than me to my right 
        for(int i=n-1;i>=0;i--){
            while(!st.empty() && heights[st.top()]>=heights[i])
                st.pop();
            if(st.empty()) left[i] = n;
            else left[i] = st.top();
            st.push(i);
        }
        int maxi = INT_MIN;
        for(int i=0;i<n;i++){
            int breadth = left[i]-right[i] -1;
            int area = breadth*heights[i];
            maxi = max(maxi , area);
        }
        return maxi;
    }
};